lib/asserts.nix

```diff
@@ -2,78 +2,122 @@
 
 rec {
 
-  /* Throw if pred is false, else return pred.
-     Intended to be used to augment asserts with helpful error messages.
+  /**
+    Throw if pred is false, else return pred.
+    Intended to be used to augment asserts with helpful error messages.
 
-     Example:
-       assertMsg false "nope"
-       stderr> error: nope
+    # Example
 
-       assert assertMsg ("foo" == "bar") "foo is not bar, silly"; ""
-       stderr> error: foo is not bar, silly
+    ```nix
+    assertMsg false "nope"
+    stderr> error: nope
+    assert assertMsg ("foo" == "bar") "foo is not bar, silly"; ""
+    stderr> error: foo is not bar, silly
+    ```
+
+    # Type
+
+    ```
+    assertMsg :: Bool -> String -> Bool
+    ```
+
+    # Arguments
+
+    `pred`
+
+    : Predicate that needs to succeed, otherwise `msg` is thrown
+
+    `msg`
+
+    : Message to throw in case `pred` fails
```

Comment:
I've been documenting the types next to the arguments/attributes, but the `lib` functions have the "Type" header which has this information too, although separate from the actual arguments/attributes. I like the "Type" header because it gives a good high-level shape of the function, but I'm suggesting a few improvements on the argument list to keep it closer to the docs for functions outside `lib`:
(applies to other functions too)

```suggestion
    1. `pred` (Bool)

       : Predicate that needs to succeed, otherwise `msg` is thrown

    2. `msg` (String)

       : Message to throw in case `pred` fails
```

Justification for the suggestions:
- this function takes two arguments, so the ordered list helps the reader know what argument they're looking at (e.g. it makes it easy for them to look up the 4th argument to a 5-argument function, for example).
- Let's also keep the type close to the arguments to make it easier for someone looking up a specific argument in isolation.

(In case one of these arguments were an attribute set, I already tested that using nested definition lists works in the rendered manual)

Reply:
So here the screenshot for completeness

![image](https://github.com/NixOS/nixpkgs/assets/50398876/369ea316-e514-47a3-8e07-c1f006332533)

Reply:
Can you try with the suggestion from my first comment? Note that the suggestion includes extra indentation for the lines that start with `: ...content`. Without that extra indentation, you'll hit the issue you described above.

Reply:
Ok. I just tried. Works, but only with exactly 4 spaces of indentation. 

Most people will have their tab on 2 spaces which increases the error potential here.
I am afraid that this convention is very brittle leading to many rendering issues, if people have leading 0-3 whitespace.

The definition list syntax is confusing enough for most of them, adding those special indentation rule or you get rendering issues is not what i'd like to aim for.

Reply:
This stuff has been bugging me for years now. Right now it looks like you keep the comments attached to arguments

Reply:
I think ideally `argument` documentation must happen by the user inside the doc-comment. 
As of RFC145 it is okay to document structured attributes outside of the doc-comment, such as:

```nix
/** Doc comment*/
{
  /** Doc for attribute 'foo' */
  foo
}:
 foo
```

We should come up with a nice convention, i think it could make sense to have everything should be at one place; in the doc-comment. Less magic, keep it simple.

```nix
/** 
Doc comment

# Inputs

...some nice convention to document foo?

*/
{
  foo
}:
 foo
```


Yes this has pros / cons, but one of the big pro is that it reduces technical complexity by a lot.